### DISCUSSION TITLE
GDPR dialog handling when moving from Google to Youtube

### DISCUSSION THREAD

**Discussion by peace2000** [uBlockOrigin/uAssets#18598] [General]
> peace2000 Can fix it? [Gif](https://i.imgur.com/no61JFO.gif) `https://www.google.com/search?q=black+sabbath`

Question from @dportvine

Personally, I could not find any working `trusted-set-cookie` filter set for that. Any ideas?

Should set "reject all".

**Comment by uBlock-user** [comment]
> Should set "reject all".

Click on reject all, immediately pull up devtools, switch to storage and look for the cookie/localStorage item set by that button. 

Now replicate that with the respective filter.

**Reply by peace2000** [reply]
That's what I tried but I couldn't find a working filter.

Though now as I investigated this more, it seems that I'd have to be able to set cookies for Youtube `before` I end up into that GDPR landing page.

So, when I use this filter:

`google.*##+js(trusted-set-cookie, SOCS, CAESNQgDEitib3FfaWRlbnRpdHlmcm9udGVuZHVpc2VydmVyXzIwMjMwNjE4LjA4X3AwGgJmaSACGgYIgPvIpAY, 1year)`

Then I go to cookiestorage and manually edit the domain of that cookie to `.youtube.com`, save and `then` click the Youtube icon, I can bypass the GDPR landing page.

This won't work:

`youtube.com##+js(trusted-set-cookie, SOCS, CAESNQgDEitib3FfaWRlbnRpdHlmcm9udGVuZHVpc2VydmVyXzIwMjMwNjE4LjA4X3AwGgJmaSACGgYIgPvIpAY, 1year)`

It will cause endless reloading.

But as set cookie doesn't have a function to set cookies for different domains, I can't do this with this scriptlet.

**Reply by peace2000** [reply]
I was trying to test how AG's `trusted-click-element` https://github.com/AdguardTeam/Scriptlets/blob/master/src/scriptlets/trusted-click-element.js would work for this, but I couldn't get it to work. Maybe because there are no clicking events associated to that "reject all" button as far as I understand.

What would help, would be that when visiting Google, it would be possible to pre-emptively set a "reject all" cookie for Youtube.

Something like this:

`google.*##+js(trusted-set-cookie, SOCS, CAESNQgDEitib3FfaWRlbnRpdHlmcm9udGVuZHVpc2VydmVyXzIwMjMwNjE4LjA4X3AwGgJmaSACGgYIgPvIpAY, 1year, , , , domain=.youtube.com)`

AG doesn't have a rule at all for that GDPR "landing" page.

@gorhill What you think about adding an option to set a cookie for a different domain than the user is currently vising?

**Reply by dimisa-RUAdList** [reply]
See also: https://github.com/uBlockOrigin/uBlock-issues/issues/2229

**Reply by gorhill** [reply]
> adding an option to set a cookie for a different domain than the user is currently vising

You can't do this, the `document.cookie` API does not allow this.

**Reply by krystian3w** [reply]
So Google in its JS asks the browser what cookies it has set for other subdomains, regional domains or seized portals? https://blog.mozilla.org/en/mozilla/firefox-rolls-out-total-cookie-protection-by-default-to-all-users-worldwide/ (Another issue is how they get around the mechanism with this, that any website can't check cookies globally and has to ask here)

**Comment by peace2000** [comment]
@dportvine

In case you use the latest dev build, you can try this set:

`consent.youtube.com##+js(trusted-click-element, 'h1 ~ div[class] div[class][jsaction] > div[class][data-is-touch-wrapper="true"] > button[jsaction]')`
`consent.youtube.com##+js(trusted-click-element, 'h1 ~ div[class] div[jsaction] > div[class][data-is-touch-wrapper="true"] > button[jsname="lW531d"]')`
`consent.youtube.com##+js(trusted-click-element, 'h1 ~ div[class] div[jsaction] > div[class][data-is-touch-wrapper="true"] > button[jsname="NHcovd"]')`
`consent.youtube.com##+js(trusted-click-element, 'h1 ~ div[class] form[class][action] > div[class][data-is-touch-wrapper="true"] > button[jsname="j6LnYe"]', , 130)`

I'm interested whether they work in different regions or not.

(Currently they don't work in uBO if combined into a one rule. That's why 4 rules that are meant to be used together.)

**Reply by dportvine** [reply]
Yep, it works for me. And it works for me with these two:

```adb
consent.youtube.com##+js(trusted-click-element, 'h1 ~ div[class] div[class][jsaction] > div[class][data-is-touch-wrapper="true"] > button[jsaction]')
consent.youtube.com##+js(trusted-click-element, 'h1 ~ div[class] form[class][action] > div[class][data-is-touch-wrapper="true"] > button[jsname="j6LnYe"]', , 130)
```

Also, in logger i see these rules without `' '`.

**Reply by peace2000** [reply]
@dportvine Yeah `' '` aren't necessary this time as there are no commas.

That dialog can be handled via 1 click if wanted, but I'm looking for a solution where uBO at first clicks "More options", then sets Youtube history "On" and Personalized ads "Off". Then confirm.

We had a `trusted-set-cookie` rule that handles the GDPR dialog on Youtube (when you visit YT directly, not from Google), but we had reports from users that uBO breaks their Youtube history so we had to adjust it so that YT history will be kept, but personalized ads not.

**Reply by dimisa-RUAdList** [reply]
I'm under the impression that the [matching feature](https://github.com/AdguardTeam/Scriptlets/blob/master/wiki/about-trusted-scriptlets.md#trusted-click-element) doesn't work as expected, at least in the case of cookie and localStorage.

But this still needs additional checking.

**Reply by ghost** [reply]
> I'm under the impression that the [matching feature](https://github.com/AdguardTeam/Scriptlets/blob/master/wiki/about-trusted-scriptlets.md#trusted-click-element) doesn't work as expected



It's not implemented 


    if ( extraMatch !== '' ) { return; }

**Reply by dimisa-RUAdList** [reply]
Yes indeed. It's a pity. In its current form, the scriptlet is sorely lacking in flexibility.

**Reply by ghost** [reply]
https://github.com/gorhill/uBlock/commit/2e4525fe3c509a636fc67776cd728819f54a7de6#commitcomment-130206560 ?

**Reply by krystian3w** [reply]
How queryjump can cover `consent.*.com`? I don't think it would allow you to load identical button layouts when Europe has 3 and the rest of the world has 2.

Perhaps, given the complexity of Google, someone will experience a redirect loop instead of a deleted cookie wall.

**Reply by peace2000** [reply]
EDIT, update 13.11.2023
It's better and quicker just to click "Reject all" to quickly get past through the dialog. We already have a `trusted-set-cookie` filter for Youtube, that will enable video recommendations and override "Reject all" cookie, so no one will complain about broken YT recommendations page: `https://www.youtube.com/feed/history`

https://github.com/uBlockOrigin/uAssets/commit/f9ee52c3a14ed555119cf0e8cfefd7a63d868723

**Reply by dportvine** [reply]
It works for me, but I don't have a YouTube account.

**Reply by garry-ut99** [reply]
> peace2000 : https://github.com/uBlockOrigin/uAssets/commit/f9ee52c3a14ed555119cf0e8cfefd7a63d868723 :
>
> https://github.com/uBlockOrigin/uAssets/blob/ad100d7bc353307c5b0bc85adf0cdeb387c8bae9/filters/annoyances-cookies.txt#L450-L451

That filter doesn't work for me (I don't know how for others), because value of `jsname` is different for me.

( To reproduce visually, don't forget to whitelist the corresponding cosmetic filter:
`consent.youtube.com#@#[data-p*="/consent.youtube.com"]` ) )

<hr>

> krystian3w : https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-1950413201 : Maybe breakage (...)

Rather not, because the consent is already clicked and hidden by `trusted-click-element`, hence hiding it additionaly by a cosmetic filter won't make a difference, but still will help to fix blinking as explained in https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-1890981370 (by the way `trusted-click-element` can click hidden elements, hence this is not a problem, even if the cookie consent will be hidden faster by a cosmetic filter, than a scriplet, the scriplet still will have not problem to click a button).

**Reply by peace2000** [reply]
@garry-ut99 adjusted: https://github.com/uBlockOrigin/uAssets/commit/e40929906c693bfc42c7f5a35bf1baffa3c8fef3

**Reply by peace2000** [reply]
The jsname is still the same on my end. This gets complicated if values differ...

What values you have?

**Reply by peace2000** [reply]
Reject all button looks like this on my end:

![image](https://github.com/uBlockOrigin/uAssets/assets/17256841/d51e046f-8e29-476d-8686-d1c2685aa118)

**Reply by garry-ut99** [reply]
Can't tell now, because it throws 404 at the moment when visiting `consent.youtube.com`, I'll tell when the error will be gone: 
![](https://ks.pl/gfx/uploads/blog/76/strona-z-bledem-404-google.webp)

**Reply by garry-ut99** [reply]
OK, I've forgotten it always throws 404 when visiting pure domain without parameters,
to avoid 404, it's needed to click a full link from this comment: https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-1950413201
as seen I'm served with a different jsname value `wsVcxe` :

![ss_02](https://github.com/uBlockOrigin/uAssets/assets/72945564/8f8a90e6-f470-45a7-9d06-f39273bc212b)

**Reply by peace2000** [reply]
@garry-ut99 What's your region?

**Reply by garry-ut99** [reply]
Poland (PL).  - but even if switched to English lang in Firefox + UK & USA IP + clean cookies & cache + restart browser, still the same value. 
So either it gives it for PL & UK & USA or maybe it's related rather to parameters in URL, 
I wonder what is your value of jsname after visiting the link from the comment.
Also I wonder what is your link where you get the old value, so I can check whether I get your value with your link or not.
That way we can determine whether it's GEO-LANG dependent or URL-PARAM depended.
There is also possibility it's USER-AGENT or FINGERPRINT dependend.

**Reply by garry-ut99** [reply]
Also I've found another cookie consent where the current filters don't work: `https://consent.youtube.com/ml?continue=https://www.youtube.com/user/GeographyNow/?ucbcb=1&%27/**/%27=0x2F2A2A2F&SLeeP%283%29&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&gl=GB&hl=en-GB&pc=yt&uxe=23983171&src=1`

![ss_04](https://github.com/uBlockOrigin/uAssets/assets/72945564/524951b0-a0de-440b-988a-d6a8c858afc1)

**Reply by peace2000** [reply]
When I open the link mentioned here: https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-1950413201

I also get `wsVcxe`. Though that page is not the "landing page":
https://github.com/uBlockOrigin/uAssets/assets/17256841/8b9d62bf-cbfd-46e7-9853-634eedb9bf36

`wsVcxe` seems to appear only in the "Personalization settings & cookies" page:
https://github.com/uBlockOrigin/uAssets/assets/17256841/73536dd0-2f4f-4394-b9f9-1a96fe469e42

AFAIK you get to that page only if you have previously clicked "More options" in the landing page. Or is there a case when you get directly to the `wsVcxe` page?

When I open this link: `https://consent.youtube.com/ml?continue=https://www.youtube.com/user/GeographyNow/?ucbcb=1&%27/**/%27=0x2F2A2A2F&SLeeP%283%29&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&ucbcb=1&gl=GB&hl=en-GB&pc=yt&uxe=23983171&src=1`

I see the same values as well. How do you reproduce that page?

Asking questions because I'm trying to carefully review this case :).

**Reply by peace2000** [reply]
My link where I get `tWT92d` is this:

`https://consent.youtube.com/m?continue=https%3A%2F%2Fwww.youtube.com%2Fchannel%2FUCQplaZZ2mv8fo8Zd-1hQPzg%3Ffeature%3Dgws_kp_artist%26feature%3Dgws_kp_artist%26cbrd%3D1&gl=FI&m=0&pc=yt&cm=2&hl=en&src=1`

**Reply by garry-ut99** [reply]
> peace2000 : My link where I get `tWT92d` is this (...)

Same value for me, ok, so mystery solved: `tWT92d` is on actual landing page, while `wsVcxe` is after clicking "More options".

> peace2000 :  AFAIK you get to that page only if you have previously clicked "More options" in the landing page. 
> peace2000 : Or is there a case when you get directly to the `wsVcxe` page?

I only used / clicked the link in https://github.com/uBlockOrigin/uAssets/issues/20586#issuecomment-1950413201, I don't  know the genesis of the link: whether the guy has clicked "More options" or has landed directly.

<hr>

> peace2000 When I open this link: (...) I see the same values as well. How do you reproduce that page?

I don't know steps to reproduce because I found it in Google search, I've also found few more:

<details>
<summary>SPOILER</summary>

`https://consent.youtube.com/ml?continue=https://www.youtube.com/user/GeographyNow?app%3Ddesktop%26ucbcb%3D1%2526%2526%2527%252F**%252F%2527%253D0x2F2A2A2F%2526%2526SLeeP%25283%2529%26ucbcb%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1&gl=GB&hl=en-GB&cm=2&pc=yt&src=1`

`https://consent.youtube.com/ml?continue=https://www.youtube.com/user/GeographyNow?app%3Ddesktop%26ucbcb%3D1%2526%2526%2527%252F**%252F%2527%253D0x2F2A2A2F%2526%2526SLeeP%25283%2529%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1%26ucbcb%3D1%26cbrd%3D1&gl=GB&hl=en-GB&cm=2&pc=yt&src=1`

</details>

**Reply by peace2000** [reply]
Now had time to check these.

Added a filter for that "modern" page: https://github.com/uBlockOrigin/uAssets/commit/2c8621b1f55a0915229965259ac814299af664ea

I think that only the transition page should be handled, not the options page where you won't land directly.

**Comment by JohnyP36** [comment]
@peace2000  I'm using those rules 
```adblock
consent.youtube.com##+js(trusted-click-element, form[action="https://consent.youtube.com/save"] button[aria-label="Reject all"], , 2000)
www.youtube.com##+js(trusted-click-element, ytd-consent-bump-v2-lightbox > #dialog > #content .eom-buttons > .eom-button-row:nth-of-type(1) > ytd-button-renderer:nth-of-type(1) button, , 20000)
```

For background info see https://github.com/gorhill/uBlock/commit/7af88b025d.

Be aware that you can't use the `:has()` operator to select an element, which is a pity.

**Reply by stephenhawk8054** [reply]
@gorhill Can we use `xpath` inside `trusted-click-element `?

**Reply by gorhill** [reply]
You can't, that scriptlet has only access to the usual CSS selectors. Procedural cosmetic filtering is a completely different library and neither has knowledge of the other.

**Reply by krystian3w** [reply]
> Be aware that you can't use the `:has()` operator to select an element, which is a pity.

Works fine for Chrome 105+.

Also works in Firefox with experimental preference (for Nightly 119-121 can be now true as default value, they probably won't make it quickly enough to make it to the stable version) - so scriptlet is not protected with same metod like procedular filters compatible with "CSS 4".

**Reply by peace2000** [reply]
@JohnyP36

Yeah, it's fine to click "reject all" but if that kind of rule is being used in uAssets, people will be unhappy because of broken Youtube history.

**Reply by JohnyP36** [reply]
@peace2000 
> Yeah, it's fine to click "reject all" but if that kind of rule is being used in uAssets, people will be unhappy because of broken Youtube history.

I know, but this kind of rule should only be for personal use. The risks are too high.
